boto/endpoints.py

```diff
@@ -0,0 +1,415 @@
+        if region_name is None:
+            if 'partitionEndpoint' in service_data:
+                region_name = service_data['partitionEndpoint']
+            else:
+                raise NoRegionError()
+        # Attempt to resolve the exact region for this partition.
+        if region_name in service_data['endpoints']:
+            return self._resolve(
+                partition, service_name, service_data, region_name)
+        # Check to see if the endpoint provided is valid for the partition.
+        if self._region_match(partition, region_name):
+            # Use the partition endpoint if set and not regionalized.
+            partition_endpoint = service_data.get('partitionEndpoint')
+            is_regionalized = service_data.get('isRegionalized', True)
+            if partition_endpoint and not is_regionalized:
+                LOG.debug('Using partition endpoint for %s, %s: %s',
+                          service_name, region_name, partition_endpoint)
+                return self._resolve(
+                    partition, service_name, service_data, partition_endpoint)
+            LOG.debug('Creating a regex based endpoint for %s, %s',
+                      service_name, region_name)
+            return self._resolve(
+                partition, service_name, service_data, region_name)
+
+    def _region_match(self, partition, region_name):
+        if region_name in partition['regions']:
+            return True
+        if 'regionRegex' in partition:
+            return re.compile(partition['regionRegex']).match(region_name)
+        return False
+
+    def _resolve(self, partition, service_name, service_data, endpoint_name):
+        result = service_data['endpoints'].get(endpoint_name, {})
+        result['partition'] = partition['partition']
+        result['endpointName'] = endpoint_name
+        # Merge in the service defaults then the partition defaults.
+        self._merge_keys(service_data.get('defaults', {}), result)
+        self._merge_keys(partition.get('defaults', {}), result)
+        hostname = result.get('hostname', DEFAULT_URI_TEMPLATE)
+        result['hostname'] = self._expand_template(
+            partition, result['hostname'], service_name, endpoint_name)
+        if 'sslCommonName' in result:
+            result['sslCommonName'] = self._expand_template(
+                partition, result['sslCommonName'], service_name,
+                endpoint_name)
+        result['dnsSuffix'] = partition['dnsSuffix']
+        return result
+
+    def _merge_keys(self, from_data, result):
+        for key in from_data:
+            if key not in result:
+                result[key] = from_data[key]
+
+    def _expand_template(self, partition, template, service_name,
+                         endpoint_name):
+        return template.format(
+            service=service_name, region=endpoint_name,
+            dnsSuffix=partition['dnsSuffix'])
+
+
+class BotoEndpointResolver(EndpointResolver):
```

Comment:
I don't think there's much that can be made private without combining the concerns of the static endpoint builder and the endpoint resolver. There are four additional methods exposed here:

* `is_global_service`: this can be removed with no issue
* `resolve_hostname`: this can also be removed since it's a convenience method, but then we would have to perform its logic in at least two places (the endpoint builder and the region connection code). I would rather keep it, but it could be removed.
* `get_available_services`: to generate the static endpoint list we need to know what all the services are. This could be replaced with a static list.
* `get_all_available_regions`: I don't think this can be removed. We need to know what endpoints are valid to build the static list. I also use that to enforce no regions are guessed, though I could potentially do that in other ways.

I could make a `StaticEndpointResolver` that subclasses `BotoEndpointResolver` which does not use the endpoint regex and which also includes the `build` function, but it seems like sharing unnecessary concerns.